### PR TITLE
Merge Autonat-svc

### PR DESCRIPTION
**Pull request by willscott**
This PR subsumes the go-libp2p-autonat-svc repository into this one.

It builds on #53 to expose both through functional arguments.
The only functional change is that the dialer used by the autonat-svc is constructed in a bit more tricky of a way to avoid a dependency on `go-libp2p`.

Note: This PR is structured to import the history from autonat-svc. The final 4 commits are the new changes: https://github.com/libp2p/go-libp2p-autonat/pull/54/files/62f42d93464d523fccc59bb7740975301d7bc1c7..dd12f0a09f839cec420686c40e8161b0a0b0158e

Known remaining work:
* [x] Combine the service and client's `background` goroutines into a single primary event loop
* [x] Pass over external interface for which constructors / methods should no longer be public

### CODE REVIEW

**Comment by vyzo** [comment]
I think this is a mistake; it was done separately precisely to avoid the dependency dance.

**Comment by willscott** [comment]
> I think this is a mistake; it was done separately precisely to avoid the dependency dance.

Separating the motivation and the mechanics of dependencies:
I think the main motivations are
(1) less chance of two versions getting out of sync and
(2) It seems like it would be nice for the consumer to say "i would like an autonat subsystem" in a relatively succinct way and have things work.

The current split between the client - which is itself of interest to the DHT and autorelay, and the service that is outside of the libp2p Host isn't the same pattern as network / routing / discovery, and is a bit of an odd interface.

In terms of dependency resolution:
* providing an option at creation for the consumer to directly choose a dialer to be used mostly removes the only dependency the service has currently on the upper libp2p repo. The rest of the dance is [newDialerFromHost](https://github.com/libp2p/go-libp2p-autonat/blob/dd12f0a09f839cec420686c40e8161b0a0b0158e/svc.go#L63-L88). I think we could arguably simplify this to just attempt the default TCP transport and error if anything else is used without an appropriate dialer passed in.

File: options.go [hunk 1]
```diff
@@ -3,26 +3,91 @@ package autonat
...
+	// ModeClient indicates that the system should detect its own NAT status,
+	// but should not help peers.
+	ModeClient
+	// ModeServer indicates that this system should always help peers determine
+	// their NAT status regardless of its own connectivity.
+	ModeServer
+)
+
+// InMode specifies the mode of AutoNAT operation.
+func InMode(mode Mode) Option {
```
Comment (Stebalien) [inline options.go]: nit: `WithMode`.

File: svc.go [hunk 2]
```diff
@@ -7,77 +7,51 @@ import (
...
 	}
 
 	as := &AutoNATService{
-		ctx:          ctx,
-		h:            h,
-		dialer:       dialer,
-		globalReqMax: AutoNATGlobalThrottle,
-		reqs:         make(map[peer.ID]int),
+		ctx:  ctx,
+		c:    c,
```
Comment (Stebalien) [inline svc.go]: nit: can we call this `config`?

File: proto.go [hunk 3]
```diff
@@ -25,7 +22,7 @@ func newDialMessage(pi peer.AddrInfo) *pb.Message {
 	}
 
 	return msg
-)
+}
```
Comment (Stebalien) [inline proto.go]: what happened here?

File: svc.go [hunk 4]
```diff
@@ -48,9 +48,10 @@ var (
 
 // AutoNATService provides NAT autodetection services to other peers
 type AutoNATService struct {
-	ctx    context.Context
-	h      host.Host
-	dialer host.Host
+	ctx         context.Context
+	h           host.Host
+	dialer      network.Dialer
+	dialerStore peerstore.Peerstore
```
Comment (Stebalien) [inline svc.go]: We should just use the one one the host.

File: svc.go [hunk 5]
```diff
@@ -48,9 +48,10 @@ var (
 
 // AutoNATService provides NAT autodetection services to other peers
 type AutoNATService struct {
-	ctx    context.Context
-	h      host.Host
-	dialer host.Host
+	ctx         context.Context
+	h           host.Host
+	dialer      network.Dialer
```
Comment (Stebalien) [inline svc.go]: These interfaces are a bit wonky and need to be refactored. Let's just use a host.

File: svc.go (hunk 4 above)
Comment (Stebalien) [reply svc.go]: Ah, now I see. We still don't need the separate peerstore if we use the `network.Network` interface.

File: svc.go (hunk 5 above)
Comment (Stebalien) [reply svc.go]: (or `network.Network`)

**Comment by Stebalien** [comment]
@vyzo the context here is:

1. The two repos are annoying to keep in sync.
2. Importantly, the service depends on events from the client to determine if it should even turn on in the first place. Otherwise, two nodes behind the same NAT could decide that they're both publicly reachable.

---

@willscott 

So, this is a bit nasty. What if we made the host a required constructor argument (for non-client services)?

We can put tests in a separate `autonat_test` package and let _that_ package import go-libp2p (circular module dependencies are fine).

File: proto.go (hunk 3 above)
Comment (willscott) [reply proto.go]: git got confused when merging a file that existed with separate histories from the two repos

File: svc.go (hunk 5 above)
Comment (willscott) [reply svc.go]: went with just a `network.Dialer`, since the dialer interface includes `.Peerstore`.

Comment (Stebalien) [reply svc.go]: `network.Network` embeds the `network.Dialer` interface.

File: options.go [hunk 6]
```diff
@@ -20,19 +37,42 @@ var defaults = func(c *config) error {
...
+		return nil
+	}
+}
+
+// AddressGuesser allows overriding which Addresses the AutoNAT client believes
 // are "its own". Useful for testing, or for more exotic port-forwarding
 // scenarios where the host may be listening on different ports than it wants
 // to externally advertise or verify connectability on.
-func WithAddresses(addrFunc GetAddrs) Option {
+func AddressGuesser(addrFunc AddrFunc) Option {
```
Comment (Stebalien) [inline options.go]: Hm. I now hate myself, that was a bad suggestion. Can you think of a better alternative?

File: options.go [hunk 7]
```diff
@@ -20,19 +37,42 @@ var defaults = func(c *config) error {
...
 	return nil
 }
 
-// WithAddresses allows overriding which Addresses the AutoNAT client beliieves
+// EnableService specifies that AutoNAT should be allowed to run a NAT service to help
+// other peers determine their own NAT status. The provided Network should not be the
+// default network/dialer of the host passed to `New`, as the NAT system will need to
+// make parallel connections, and as such will modify both the associated peerstore
+// and terminate connections of this dialer. The dialer provided
+// should be compatible (TCP/UDP) however with the transports of the libp2p network.
```
Comment (Stebalien) [inline options.go]: This isn't strictly speaking necessary. Ideally, the dialer would support dialing _all_ available transports.

File: options.go [hunk 8]
```diff
@@ -20,19 +37,42 @@ var defaults = func(c *config) error {
...
+func EnableService(dialer network.Network, forceServer bool) Option {
+	return func(c *config) error {
+		if dialer == c.host.Network() || dialer.Peerstore() == c.host.Peerstore() {
+			return errors.New("dialer should not be that of the host")
```
Comment (Stebalien) [inline options.go]: :+1:

File: autonat.go [hunk 9]
```diff
@@ -236,6 +255,9 @@ func (as *AmbientAutoNAT) recordObservation(observation autoNATResult) {
 				// we are flipping our NATStatus, so confidence drops to 0
 				as.confidence = 0
 				as.status.Store(observation)
+				if as.serviceCancel != nil {
+					as.serviceCancel()
```
Comment (Stebalien) [inline autonat.go]: nit: might as well have `as.serviceCancel = nil`.

File: autonat.go [hunk 10]
```diff
@@ -252,13 +274,16 @@ func (as *AmbientAutoNAT) recordObservation(observation autoNATResult) {
 		log.Debugf("NAT status is unknown")
 		as.status.Store(autoNATResult{network.ReachabilityUnknown, nil})
 		if currentStatus.Reachability != network.ReachabilityUnknown {
+			if as.serviceCancel != nil {
+				as.serviceCancel()
```
Comment (Stebalien) [inline autonat.go]: Same here.

**Pull request merged**